romdisk/uart.asm

```diff
@@ -144,48 +144,140 @@ uartsnd_main:
         ; It will try to open the driver registered as SER0.
-        ; TODO: Implement this command to actually save data in a file.
         ; Parameters:
         ;       HL - ARGV
         ;       BC - ARGC
         ; Returns:
         ;       A - 0 on success
+        ; Modifies:
+        ;       IX, DE
         PUBLIC uartrcv_main
 uartrcv_main:
+        ; Set up stack frame
+        ld ix, 0
+        add ix, sp
+
         dec bc  ; Do not count the command itself
+
+        ; Check that size is present
         ld a, b
         or c
-        ret z
+        jp z, _rcv_usage
+        ; Advance past command 
         inc hl
         inc hl
+
+        ; Load stack frame
+        push hl ; remaining argv (ix-2)
+        push bc ; remaining argc at (ix-4)
+
+        ; Parse the length parameter
         ld a, (hl)
         inc hl
         ld h, (hl)
         ld l, a
-        ; Parse the length parameter
         call parse_int
-        ; Check if A is 0 or not
+        ; Check parse (a should be zero)
         or a
-        ; At the moment, simply return, this command is just for the sake of demonstration
-        ret nz
-        ; Save the buffer size
-        push hl
+        jp nz, _rcv_usage
+        ; Add read/write size to stack frame
+        push hl ; size at (ix-6)
+
+        ; Open UART
         call uart_open
-        ; Save the size in BC
-        pop bc
-        jp m, open_error
-        ; Read from the UART
-        ; DE - Buffer
-        ; BC - Size
-        ; H  - Descriptor
+        jp m, _uart_rcv_clean_and_error
+
+        ; Set up and do read
+        ld b, (ix-5)
+        ld c, (ix-6)
         ld de, 0x8000
         ld h, a
         READ()
-        ; TODO: Show what was actually received
-        ; Close the driver
+        ; Check read return
+        or a
+        jp nz, _uart_rcv_clean_and_error
+        ; Close the UART driver
+        CLOSE() 
+
+        ; Prepare for write: load H with either stdout or opened file
+        call _uart_rcv_prep_output
+        or a
+        jp nz, _uart_rcv_clean_and_error
+
+        ; Successful prep, h has open device
+        ld b, (ix-5)
+        ld c, (ix-6)
+        ld de, 0x8000
+        WRITE()
+        ; If output was not DEV_STDOUT, close it (needed to flush file output)
+        ld a,h
+        cp DEV_STDOUT
+        jr z, _uart_rcv_clean_and_return
         CLOSE()
+
+_uart_rcv_clean_and_return:
+        ld sp,ix
         ret
 
+_uart_rcv_prep_output:
+        ; Return either stdout or opened output file, depending on whether
+        ; there is a third parameter (second now that we have advanced past
+        ; the command name).
+        ; Parameters:
+        ;   (IX-2) remaining argv
+        ;   (IX-4) remaining argc
+        ; Returns:
+        ;   A - zero on success, open error on failure
+        ;   H - descriptor on success
+        ; Modifies: 
+        ;   BC
+
+        ; Check if remaining argc > 1
+        ld  b, (ix-3)
```

Comment:
If you parse the file name at the routine entry, you can actually take that address as a parameter here, it would make your routine smaller and faster

Reply:
See new commit.